### PR TITLE
[BEAM-2899][BEAM-3329][BEAM-3328] Link up the Portable DirectRunner

### PR DESCRIPTION
**Pull request by tgroh**
This can execute pipelines via the Portability APIs, but here only via
the InProcess channel.

Interface and dead code cleanups are also included here.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

### CODE REVIEW

**Comment by tgroh** [comment]
CC: @herohde

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/PortableDirectRunner.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+/** The "ReferenceRunner" engine implementation. */
+class PortableDirectRunner {
+  private final RunnerApi.Pipeline pipeline;
+
+  private PortableDirectRunner(RunnerApi.Pipeline p) throws IOException {
+    this.pipeline = executable(p);
+  }
+
+  static PortableDirectRunner forPipeline(RunnerApi.Pipeline p) throws IOException {
```
Comment (lukecwik) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/PortableDirectRunner.java]: Shouldn't we take the PipelineOptions here and use it down below?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectStateAndTimers.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+package org.apache.beam.runners.direct.portable;
+
+import org.apache.beam.runners.direct.Clock;
+import org.apache.beam.runners.direct.WatermarkManager.TimerUpdate;
+import org.apache.beam.runners.direct.WatermarkManager.TransformWatermarks;
+import org.apache.beam.runners.local.StructuralKey;
+
+/**
+ * Execution Context for the {@code DirectRunner}.
```
Comment (lukecwik) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectStateAndTimers.java]: Why do you want to move away from using this as an execution context?
(I don't see the big picture of how the pieces will fit together.)

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/portable/PortableDirectRunnerTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                            KV.of(
+                                ctxt.element().getKey(),
+                                ImmutableSet.copyOf(ctxt.element().getValue())));
+                      }
+                    }));
+
+    p.replaceAll(Collections.singletonList(JavaReadViaImpulse.boundedOverride()));
+
+    PortableDirectRunner runner = PortableDirectRunner.forPipeline(PipelineTranslation.toProto(p));
+    runner.execute();
```
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/portable/PortableDirectRunnerTest.java]: Sorry, that was incorrect.

Currently, adding a PAssert causes the pipeline to get wedged somewhere in the Fn Harness. I don't know why, but I'm investigating.

In the meantime, I've manually verified that the result is as expected. I'd prefer to include a TODO that this is under investigation and continue making progress on getting the RR into `master` than block submissions on this, assuming that you're OK with that.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/PortableDirectRunner.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+import org.apache.beam.runners.fnexecution.environment.InProcessEnvironmentFactory;
+import org.apache.beam.runners.fnexecution.logging.GrpcLoggingService;
+import org.apache.beam.runners.fnexecution.logging.Slf4jLogWriter;
+import org.apache.beam.runners.fnexecution.state.GrpcStateService;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** The "ReferenceRunner" engine implementation. */
+class PortableDirectRunner {
```
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/PortableDirectRunner.java]: `ReferenceRunner`, wheeeeee

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/portable/PortableDirectRunnerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/portable/PortableDirectRunnerTest.java]: Fixed the actual bug.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectStateAndTimers.java (hunk 2 above)
Comment (lukecwik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectStateAndTimers.java]: SGTM

Comment (lukecwik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectStateAndTimers.java]: Can you update the comment to reflect the narrow scope and intention of this class.

**Pull request merged**